### ISSUE TITLE
cyrillic breaks in questions with an image

### ISSUE THREAD

**Issue by lebedevalex3** [maths/moodle-qtype_stack#995]
Hello everyone! After the STACK update, cyrillic breaks in questions with an image. The bug has appeared in all old questions and appears in new ones.

With image:
![image](https://github.com/maths/moodle-qtype_stack/assets/109303316/bd0b3b6c-884d-459e-a66e-bf2e8acc3b9e)
without image:
![image](https://github.com/maths/moodle-qtype_stack/assets/109303316/51a4fa4a-6d7b-40e0-8dab-86892d89f85e)

**Comment by sangwinc** [comment]
Oh dear, that's odd.  Can you confirm this problem does not occur if you put just the text and the image into Moodle's "description" question type?  This will confirm it probably is a "STACK thing".

If so, please email me a question and I'll take a look.

Chris

**Comment by lebedevalex3** [comment]
In other questions, everything works fine.
![image](https://github.com/maths/moodle-qtype_stack/assets/109303316/278d8dbe-4824-4af1-9a69-81b630e496dc)

I installed the FreeSerif font, but even that didn't help. Then I rolled back and installed the previous version of STACK - everything works. Somehow, embedding an image affects the encoding. I'm using Moodle 4.2.

**Comment by lebedevalex3** [comment]
I have created a test website (Ubuntu 20) and installed Moodle 4.2. Maxima 5.46.0 is also installed. The behavior is the same - everything works fine when there is no image, but when an image appears, the font changes to question marks.


![image](https://github.com/maths/moodle-qtype_stack/assets/109303316/cea127ba-02dc-42c7-a6b4-44604071b825)


You can observe this strange behavior on the website:
http://185.233.187.102/
Login: admin
Password: P123-test

If needed, I can provide access to the server.

**Comment by sangwinc** [comment]
Thank you for providing such a helpful report of the problem, and a link to the question. I've reproduced the problem on your server, obviously!  I've also downloaded the question onto my own server.  I can't reproduce the problem on my server which suggestions there is something on your moodle setup which is significantly contributing to the problem.  That's going to make it harder to track down.

I don't have any colleagues from Russia on the STACK Professionals network (yet) https://stack-assessment.org/Community/ who might be able to offer advice, and either confirm that STACK works or confirm your problem is more general.  However, I don't have lots of bug reports from other multi-language users (many are in Hebrew, Japanese, etc. using non-UK character sets).

1. I checked your Moodle filter settings, and there are no differences from mine.
2. You are using PHP 8, whereas my dev server is still on PHP 7.  We have unit test coverage of PHP 8, but failure to show fonts would not be caught by that kind of testing probably.

Honestly, I'm really not sure what is wrong or what to test next!  I'll have to give this some more thought.  

@timhunt, @aharjula  have you seen this problem before?

**Comment by sangwinc** [comment]
I also tried the following.

1. Put mathjax filter first.
2. Turn off all filters except mathjax

Neither approach fixed the problem.  I've done my best to restore your filter settings to their original values (which matched the defaults on my dev server to start with).

**Comment by aharjula** [comment]
Probably broken anywhere where the CASText has any depth that is not being handled by the static strings logic, i.e., this is likely an issue with Maxima having a bad unicode setup. For example, looking into `/question/type/stack/adminui/compiledquestion.php?qid=4`, one can see that the text is going to be sent to the CAS with the chars inline (mainly because, for whatever reason, the static string extraction does not act on the PFS (pluginfile) blocks contents).

That theory should be easy to test by trying a CASText fragment with very short sections of chars and some injection between them and seeing if that causes this same thing. For example, `Да {@"Да"@} Да`, which will probably generate plenty of question marks. Short sections are any sections shorter than the keys the static string logic generates and thus not being processed by it.

**Comment by sangwinc** [comment]
Thanks Matti,

Yes, Maxima on your system does not support unicode.  The Healthcheck script gives the error "Maxima encountered a Lisp error: WRITE-CHAR: Character #\u2200 cannot be represented in the character set CHARSET:ASCII Automatically continuing."
This is a problem more generally.

Your example also breaks Matti:  `Да {@"Да"@} Да` gives no meaningful output, just ?s.

What I don't understand is why this example breaks:  `Да  {@plot(x^2,[x,-2,2])@}`.   

It is strange that the static string extraction does not act on the PFS (pluginfile) blocks contents, and on my minimal example.  However the system will break without unicode support for Maxima quite quickly anyway when you do send content there, so perhaps breaking everything is a useful warning there is an underlying problem.

**Comment by aharjula** [comment]
Chris, it breaks for the very same reason as mine. The fragment before the injection of the plot is so small that it gets passed into the CAS raw and will not survive if the CAS is not Unicode capable.

It does not act on PFS contents because someone has not remembered to make it act on them. Basically, as blocks can encode themselves into anything they like, the static extraction needs to be told which parts are safe for extraction and which parts might be needed in raw form on the CAS side, in general almost all parts where we have string atoms at the list level should be safe excluding the name of the block.

Basically, adding the name of the PFS block [here](https://github.com/maths/moodle-qtype_stack/blob/ec5cc8e6dffb2b915c2fa682870e2c1ef8af396a/stack/cas/parsingrules/610_castext_static_string_extractor.filter.php#L47) would move this problem elsewhere for larger fragments.

One could also think about whether all short fragments should be considered for extraction if they contain non-ASCII chars. But then the issue of broken Maxima config would be hidden...

**Comment by lebedevalex3** [comment]
Thank you all for the help!
The maxima build was done using **clisp**. After the build was done through **sbcl**, the problem went away.
Thanks again!
